source/Random.cpp

```diff
@@ -104,12 +111,33 @@ uint32_t Random::Binomial(uint32_t t, double p)
 
 
 
-// Get a normally distributed number (mean = 0, sigma= 1).
-double Random::Normal()
+// Get a normally distributed number with default or specified mean and stddev using the Box-Muller transform.
+// Cache the unused value without transforming it so that it can be transformed when it's used.
+double Random::BoxMullerSample(double mean, double sigma)
 {
-	normal_distribution<double> normal;
-#ifndef __linux__
-	lock_guard<mutex> lock(workaroundMutex);
-#endif
-	return normal(gen);
+	if(normalBMCached)
+	{
+		#ifndef __linux__
+			lock_guard<mutex> lock(workaroundMutex);
+		#endif
+		normalBMCached = false;
```

Comment:
You must lock _before_ you read `normalBMCached`. Otherwise two threads can still enter the if block at the same time, both seeing a value of `true`, both setting it to false, and both using the cached value.

The reason you obtained a hang is because you did not release the lock before calling `Random::Real()`, causing a deadlock where you attempted to acquire the same lock again.

Reply:
Would this be more correct: c9218fc09919de4ff4f555a56f8459984be15056?